cEP-0018.md

```diff
@@ -0,0 +1,164 @@
+| Title    | Integration of ANTLR into coala core          |
+| Authors  | Viresh Gupta <mailto:viresh16118@iiitd.ac.in> |
+| Status   | Proposed                                      |
+| Type     | Feature                                       |
+
+## Abstract
+
+This document describes how an API based on ANTLR will be constructed and
+maintained.
+
+## Introduction
+
+ANTLR provides parsers for various language grammars and thus we can provide
+an interface from within coala and make it available to bear writers so that
+they can write advanced linting bears. This will aim at supporting the more
+flexible visitor based method of AST traversal (as opposed to listener based
+mechanisms).
+
+The proposal is to introduce a parallel concept to the coala-bears library,
+which will be called the coala-ast hereon.
+
+## Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. There will be a separate repository named as coala-antlr-ast which will
+   be installable via `pip install coala-antlr-ast`.
+2. A new package would be introduced in the coala-antlr-ast repository,
+   which would be the maintain all the dependencies for antlr-ast based
+   bears. This would be the `antlrastlib` package
+3. The `antlrastlib` package will provide endpoints relevant to the
+   visitor model of traversing the ast generated via ANTLR.
+4. Another package named `antlrparsers` will be created inside
+   coala-antlr-ast repository. This package will hold parsers generated
+   for various languages generated (with python target) beforehand for the
+   supported set of grammars.
+5. `antlrastlib` will have an AST loader class that will be responsible
+   for loading the AST of a given file depending on the extension.
+   For e.g, a .c file will be loaded using the parser from
+   `antlrastlib` for c language.
+6. Another AST walker class will be responsible for providing endpoints for
+   traversing the AST using methods such as `get_next_node` and
+   `get_previous_node`.
+
+## Management of the new `coala-ast` repository
+
+Managing a new repository is a heavy task, and this will be highly automated as
+follows:
+
+1. The parsers in coala-antlr-ast/antlrparsers would be generated and
+   pushed via travis-builds whenever a new commit is pushed.
+2. The cib tool can be enhanced to deal with the installation of bears that
+   require only some specified parsers (for e.g a PyUnusedVarBear would
+   only require parser for python)
+3. The cib tool can also trigger specialised builds and download the newly
+   generated parser on the fly
+
+## Code Samples/Prototypes
+
+Here is a prototype for the implementations within antlrastlib:
+
+```python
+import antlrparsers
+import antlr4
+
+class ASTLoader():
+    lexerMapping = {
+        'c' : antlrparsers.clexer,
+        'py' : antlrparsers.pylexer,
+        'js' : antlrparsers.jslexer,
+        ...
+    }
+    parserMapping = {
+        'c' : antlrparsers.cparser,
```

Comment:
is there always a parser for every lexer?  1:1 mapping.  If so, two dicts isnt a good way to map that into Python.

Reply:
Yep,
We have to have a lexer for every parser otherwise we wouldn't be able to tokenize the source code to feed into the parser

Probably one dictionary with an ordered tuple then ?